### PR TITLE
CQRS-2/Create 3 Producers

### PR DESCRIPTION
**Pull request by shreyk2**

### CODE REVIEW

File: service-producer/pom.xml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		<connection/>
+		<developerConnection/>
+		<tag/>
+		<url/>
+	</scm>
+	<properties>
+		<java.version>21</java.version>
+		<spring-cloud.version>2023.0.2</spring-cloud.version>
+	</properties>
+	<dependencies>
```
Comment (Hannibal-Tao) [inline service-producer/pom.xml:33]: I think you can set up the profiles for local and docker environments
for example:
```
<profiles>
    <profile>
      <id>local</id>
        <activation>
          <activeByDefault>true</activeByDefault>
        </activation>
      <properties>
        <spring.profiles.active>local</spring.profiles.active>
        <environment>local</environment>
      </properties>
    </profile>
    <profile>
      <id>docker</id>
      <properties>
        <spring.profiles.active>docker</spring.profiles.active>
        <environment>docker</environment>
      </properties>
    </profile>
  </profiles>
```

File: service-producer/pom.xml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
+		</dependency>
+<!--		<dependency>-->
```
Comment (Hannibal-Tao) [inline service-producer/pom.xml]: You can uncomment from the 38th line to the 45th line.

File: service-producer/src/main/java/com/cqrs/ServiceProducer1Application.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.cqrs;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+//import org.springframework.cloud.client.discovery.EnableDiscoveryClient;
+
+//@EnableDiscoveryClient
```
Comment (Hannibal-Tao) [inline service-producer/src/main/java/com/cqrs/ServiceProducer1Application.java]: uncomment this line to turn on the zookeeper.

File: service-producer/src/main/resources/application.properties [hunk 4]
```diff
@@ -0,0 +1,3 @@
+spring.application.name=service-producer-1
+spring.cloud.zookeeper.connect-string=172.18.0.4:2181,172.18.0.2:2181,172.18.0.3:2181
```
Comment (Hannibal-Tao) [inline service-producer/src/main/resources/application.properties]: To utilize the profile feature of Spring for separating environment properties, you can move the second line into different property files.
for example
add 2 property files, application-local.properties, and application-docker.properties.
the local one is as follows
```
spring.cloud.zookeeper.connect-string=localhost:2181,localhost:2182,localhost:2183
```
the docker one is as follows
```
spring.cloud.zookeeper.connect-string=zookeeper-1:2181,zookeeper-2:2181,zookeeper-3:2181
```
when packaging to a jar file using Maven,
you can pass the profile value to specify the environment explicitly.
for example, if you want to build a jar for the docker environment.
```
./mvnw clean package -P docker
```

File: service-producer/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,5 @@
+FROM openjdk:21-jdk
```
Comment (Hannibal-Tao) [inline service-producer/Dockerfile:9]: you can try using a builder image
https://docs.docker.com/build/building/multi-stage/

**Comment by shreyk2** [comment]
Completed the review changes

**Review by Hannibal-Tao — APPROVED** [review]
Looks good

**Pull request merged**